src/providers/crypto/cryptors/aes_cbc.rs

```diff
@@ -0,0 +1,205 @@
+//! # AES-CBC cryptor module.
+//!
+//! Module contains [`AesCbcCryptor`] type which can be used for data encryption
+//! and decryption.
+
+use aes::cipher::{block_padding::Pkcs7, BlockDecryptMut, BlockEncryptMut, KeyIvInit};
+use sha2::{Digest, Sha256};
+
+use crate::{
+    core::{Cryptor, EncryptedData, PubNubError},
+    lib::alloc::{format, string::ToString, vec, vec::Vec},
+};
+
+type Encryptor = cbc::Encryptor<aes::Aes256>;
+/// AES-SHA256 encryptor type.
+type Decryptor = cbc::Decryptor<aes::Aes256>;
+
+/// Unique cryptor identifier
+const IDENTIFIER: [u8; 4] = *b"ACRH";
+
+/// AES cipher block size.
+const AES_BLOCK_SIZE: usize = 16;
+
+/// AES-CBC cryptor.
+#[derive(Debug)]
+pub struct AesCbcCryptor {
+    /// Key for data _encryption_ / _decryption_.
+    cipher_key: Vec<u8>,
+}
+
+impl AesCbcCryptor {
+    /// Create AES-CBC cryptor.
+    pub fn new<K>(cipher_key: K) -> Result<Self, PubNubError>
+    where
+        K: Into<Vec<u8>>,
+    {
+        let cipher_key = cipher_key.into();
+
+        if cipher_key.is_empty() {
+            return Err(PubNubError::CryptoInitialization {
+                details: "Cipher key is empty".into(),
+            });
+        }
+
+        Ok(Self {
+            cipher_key: Self::sha256(cipher_key),
+        })
+    }
+
+    fn initialization_vector(&self) -> [u8; 16] {
+        let mut random = [0u8; AES_BLOCK_SIZE];
+        getrandom::getrandom(&mut random).ok();
```

Comment:
I'm actually not sure about this.  The `getrandom` crate's docs says this:

> It is assumed that the system always provides high-quality cryptographically secure random data

Is it safe to make this assumption?  On the other hand, even the `rand` crate, which explicitly says it provides cryptographically secure random number generators, says this:

> A Rust library for random number generation, featuring secure seeding via the getrandom crate

So it isn't clear that using the `rand` crate would be any better.  My best guess is that this will be fine for nearly all customers.  However, there may be customers whose software runs in an environment (e.g. embedded hardware) that does not have access to cryptographically secure seeding.  While it would be possible to allow customers to provide their own custom RNG in those cases, probably it's best that those customers just don't use this feature.

All of which is to say I am not recommending any change here.

Reply:
Yes. It looks like all suitable (`no_std`) CSPRNGs from [this](https://rust-random.github.io/book/guide-rngs.html#cryptographically-secure-pseudo-random-number-generators-csprngs) list rely on `getrandom` when we need to create an instance of random number generator by requesting create it from `entropy` – this under the hood relies on the `getrandom` feature.

Currently, it is hard to say what is the best solution here. With future updates, _maybe_ we will create some deviation (from other SDKs) in the interface to let user specify closure which will generate random data for IV (and salt when we will switch to KDF).